docs: forward-ports from 3.x

tchaikov opened the pull request:
the document on 4.x branch does not build without this changeset. in order to build docs from 4.x branch, we have to have https://github.com/scylladb/java-driver/commit/cdaf75a6cf54b91f3f795aa509cf836166dad1e2. but since we are here, let's forward-port all 3.x doc change to 4.x. 

tested using 

```console
$ make setup
$ make preview
```

avelanarius committed:
Pin specific version of Jinja2

Before this change, the docs failed to build with the following
error:

ImportError: cannot import name 'environmentfilter' from 'jinja2'

A fix to this problem is to explicitly pin Jinja2 version.

(cherry picked from commit cdaf75a6cf54b91f3f795aa509cf836166dad1e2)

dgarcia360 committed:
docs: update theme 1.2

docs: update dynamic slug

Add back index.md

Move custom slug to a separate PR

Fix typo

(cherry picked from commit 8feeb205fcf163d9a45dbe884fc353ef5a1a4535)

avelanarius committed:
Add Java Driver 4.x branches to documentation

Add Java Driver 4.x branches to the list of generated documentation
and bump the LATEST_VERSION variable.

(cherry picked from commit 76f771528db530c8af7f38551869fcf0d13ce567)

dgarcia360 committed:
docs: update theme 1.3

(cherry picked from commit d2e261bcf437bd4863ce22215fdf5a4bfbf1f7de)

dgarcia360 committed:
fix: redirection to api reference

Fix javadoc redirect

Fix

(cherry picked from commit a9fcffb44873eb938eb0312106d0c83de8f8e4dd)

dgarcia360 committed:
Add license notice

(cherry picked from commit 7399a0ad40999f203639330d88ebaf99db6a0d10)

avelanarius committed:
Add java-driver.docs.scylladb.com replacements

The links in our documentation have a hardcoded version number in them:

https://java-driver.docs.scylladb.com/scylla-3.11.2.x/api/
https://docs.datastax.com/en/drivers/java/3.11/com/datastax/driver/core/Cluster.html

Therefore, a replacement rule is needed to replace the version to a
correct version for a specific driver version we render the docs for.

Before this change, only datastax links were replaced. After this
change, java-driver.docs.scylladb.com links are also replaced.

(cherry picked from commit 8cd4d8e162226419b3d85babb57d4db7bd0ea2c1)

avelanarius committed:
Fix java-driver.docs.scylladb.com replacement

(cherry picked from commit 511573f814770cb4faa2e7136e1d046d6f196729)

dgarcia360 committed:
docs: update theme

docs: update conf.py

docs: update conf.py

docs: Update conf.py

docs: update conf.py

docs: Update conf.py

docs: update conf.py
(cherry picked from commit 0d1f1c4512afe7bc48fec373d41f802ca17e74de)

dgarcia360 committed:
docs: update conf.py

docs: test

docs: add skip warnings option
(cherry picked from commit 2601b782e339ab0725858346918699f3da913732)

dgarcia360 committed:
docs: update theme 1.5

(cherry picked from commit 51ff0d5988270027e80a5d428ec0ce711619fb61)

avelanarius committed:
Add 4.14.1.x and 4.15.0.x to generated docs

Add 4.14.1.x and 4.15.0.x releases to the list of versions to generate
the docs for.

(cherry picked from commit d29689596daf2f9d8002e5f71c743a088b1b95a2)

tchaikov committed:
docs: wrap long line for better readability

Signed-off-by: Kefu Chai <kefu.chai@scylladb.com>
(cherry picked from commit 6a5d3a7f9d37cc9f25f184d5b90169bb0901fc3d)

tchaikov commented:
/cc @dgarcia360 @annastuchlik

dgarcia360 committed:
docs: update theme 1.6

(cherry picked from commit 7ab2a8f90b1e9bb7927aecb5a0cbe5cb629b9202)

dgarcia360 committed:
docs: update workflows

(cherry picked from commit 8dea9e264368827bd507efd641f3d626eaf58258)

roydahan commented:
Need to rebase PR in order to re-run the CI.

tchaikov commented:
rebased and repushed.

avelanarius commented:
@dgarcia360 @annastuchlik Please review this PR.

@tchaikov I'm not really sure I understand the reasoning for this PR. The cover letter doesn't really explain the intention.

tchaikov commented:
> @dgarcia360 @annastuchlik Please review this PR.
> 
> @tchaikov I'm not really sure I understand the reasoning for this PR. The cover letter doesn't really explain the intention.

@avelanarius hi Piotr, the document on 4.x branch does not build without this changeset. that's why i ported the change to address the FTBFS.

dgarcia360 commented:
@annastuchlik @tchaikov @avelanarius  It looks good to me. The issue is that, without these changes, the 4.x will not build locally.

dgarcia360 committed:
docs: add myst parser

(cherry picked from commit 9eedcaca09d1bb5bb87452389c1eb9cc1bd9645f)

dgarcia360 committed:
docs: update theme

(cherry picked from commit efe510181a8c6a105f0ad953d9427ce4550eb671)

roydahan commented:
Ping reviewers

Bouncheck commented on .github/workflows/docs-pages.yaml:
```diff
@@ -0,0 +1,43 @@
+on:
+  push:
+    branches:
+    - scylla-3.x
```
This feels wrong. Wouldn't that never fire?

Bouncheck commented on .github/workflows/docs-pr.yaml:
```diff
@@ -0,0 +1,36 @@
+on:
+  pull_request:
+    branches:
+    - scylla-3.x
```
Similar as above. We should not have a workflow for PRs to 3.x on 4.x branch.

Bouncheck reviewed (changes requested):
I successfully ran `make preview` after checking out this branch, however I've seen that script had to downgrade a lot of dependencies. Maybe a rebase is needed.
Github workflows seem incorrect or unnecessary: do we do docs deploy from 4.x branch? Deploy.sh seems not to be there anyway.
Is this purely a port? CustomCommonMarkParser seems to be new, do we not need it in 3.x?
Branchest/Tags need an update: latest tag is 4.18.0.1

Bouncheck commented:
How do we maintain those two docs branches going forward? Is it possible to make it so we keep up to date setup on `scylla-3.x` and from time to time copy whole `docs` subdirectory to `scylla-4.x`?

tchaikov committed:
docs: triggered when on push  to scylla-3.x

this change is not cherry-picked from 4.x, as it is specific to
scylla-3.x branch.

Signed-off-by: Kefu Chai <kefu.chai@scylladb.com>

tchaikov commented:
> I successfully ran `make preview` after checking out this branch, however I've seen that script had to downgrade a lot of dependencies. Maybe a rebase is needed. 

sure. rebased.

> Github workflows seem incorrect or unnecessary: do we do docs deploy from 4.x branch? 

i didn't manually edit the backport commits unless i had to to resolve the conflicts or to fix the document build. the purpose of this changeset is to enable us to build document on 3.x branch. i am not sure if we deploy docs from 4.x branch. guess this is a question to the maintainer of this project.

considering this PR has been around for half a year, can we get this changeset reviewed and then work on the CI of document deployment if this is desirable?

> Deploy.sh seems not to be there anyway. Is this purely a port? 

yes, it is a backport. but again, this PR has been around for ages. and i haven't cherry-picked new changes from 4.x since i sent it for review. if something is odd. the reason is very likely that it's caused by merge conflict or missing commits from 4.x. 

> CustomCommonMarkParser seems to be new, do we not need it in 3.x? 

on the contrary, it is old. the change removing `CustomCommonMarkParser` landed in 9eedcaca09d1bb5bb87452389c1eb9cc1bd9645f 

> Branchest/Tags need an update: latest tag is 4.18.0.1

could you be more specific? which files shall i update?

tchaikov commented:
v2:

- cherry-pick more changes to `docs/source/conf.py`  from 4.x 
- suppress warning from autosectionlabel
- correct the codeblock directive (the patch was also sent to upstream, see https://github.com/apache/cassandra-java-driver/pull/1940)

tchaikov commented:
> How do we maintain those two docs branches going forward?

again, i think it's up to the maintainer of this project.

>  Is it possible to make it so we keep up to date setup on `scylla-3.x` and from time to time copy whole `docs` subdirectory to `scylla-4.x`?

i don't think to "copy" the whole `docs` subdirectory to `scylla-4.x` branch is the right approach. my 2 cents would be

1. get the 4.x doc build in shape, so that it can at least build
2. decide if / how to deploy it
3. setup the CI to backport and deploy it
4. add the proper label when sending doc patches to 3.x, and let CI to backport the change.

tchaikov committed:
docs: suppress warning from autosectionlabel

to silence the warning from sphinx, as we have two "Dropwizard
Metrics" sections in manual/core/metrics/index.md:

```
/home/kefu/dev/scylla-java-driver/docs/_source/manual/core/metrics/index.md:291: WARNING: duplicate label manual/core/metrics/index:dropwizard metrics, other instance in /home/kefu/dev/scyll
a-java-driver/docs/_source/manual/core/metrics/index.md
```

and the warning breaks the "make test". so let's suppress the warning
at this moment.

Signed-off-by: Kefu Chai <kefu.chai@scylladb.com>

tchaikov committed:
manual: correct the codeblock directive

it should start with three backticks not four of them. otherwise
sphinx warns like:

```
/home/kefu/dev/scylla-java-driver/docs/_source/manual/mapper/daos/getentity/index.md:111: WARNING: Lexing literal_block '@GetEntity\nProduct asProduct(Row row);\n\n@GetEntity\nProduct firstR
owAsProduct(ResultSet resultSet);\n```\n\n' as "java" resulted in an error at token: '`'. Retrying in relaxed mode.
```

in this change, let's use three backticks. and the warning disappears.

Signed-off-by: Kefu Chai <kefu.chai@scylladb.com>

tchaikov commented:
@Bouncheck hi Wojciech, could you please take another look?

Bouncheck commented:
`make clean preview` looks better. Now manual pages can be browsed normally (before they did not work). API Documentation does not link to javadoc.

`make clean multiversionpreview` does not build correctly for me. For testing I've removed all but 2 versions (4.14.1.x and 4.15.0.x) from `BRANCHES` in source/conf.py to speed up the build.  I'm getting a 404 error page instead.
Throughout the build I'm getting a lot of lines like this one, after "updating environment" step:
```
updating environment: [new config] 102 added, 0 changed, 0 removed
/home/wo/.cache/pypoetry/virtualenvs/java-driver-xPfC9vS5-py3.11/lib/python3.11/site-packages/recommonmark/parser.py:75: UserWarning: Container node skipped: type=document
  warn("Container node skipped: type={0}".format(mdnode.t))
/home/wo/.cache/pypoetry/virtualenvs/java-driver-xPfC9vS5-py3.11/lib/python3.11/site-packages/recommonmark/parser.py:75: UserWarning: Container node skipped: type=document
  warn("Container node skipped: type={0}".format(mdnode.t))
```
Visiting the `127.0.0.1:5500` directs me to `http://127.0.0.1:5500/master` which has 404 error. However pages like `http://127.0.0.1:5500/stable/index.html` or `http://127.0.0.1:5500/scylla-4.15.0.x/` remain functional with redirects to javadoc when clicking on API Documentation header too.

> >    Branchest/Tags need an update: latest tag is 4.18.0.1
>
>could you be more specific? which files shall i update?

I was guessing that we'd need to just add '4.18.0.1' to `TAGS` in docs/source/conf.py but that actually breaks the build with multiple warnings like:
```
/tmp/tmpk4aq9ak4/970452ef5b72d76b4f783335fccabe57107d9735/docs/source/manual/core/pooling/index.md: WARNING: document isn't included in any toctree
```
So after all let's not do that for now.

tchaikov commented:
> `make clean preview` looks better. Now manual pages can be browsed normally (before they did not work). API Documentation does not link to javadoc.
> 
> `make clean multiversionpreview` does not build correctly for me. For testing I've removed all but 2 versions (4.14.1.x and 4.15.0.x) from `BRANCHES` in source/conf.py to speed up the build. I'm getting a 404 error page instead. Throughout the build I'm getting a lot of lines like this one, after "updating environment" step:
> 

i commented out "multiversionpreview" to workaround a dependency issue on python 3.12, but forgot to revert it. the latest version dropped this workaround. so the multiversion should be back now.

> ```
> updating environment: [new config] 102 added, 0 changed, 0 removed
> /home/wo/.cache/pypoetry/virtualenvs/java-driver-xPfC9vS5-py3.11/lib/python3.11/site-packages/recommonmark/parser.py:75: UserWarning: Container node skipped: type=document
>   warn("Container node skipped: type={0}".format(mdnode.t))
> /home/wo/.cache/pypoetry/virtualenvs/java-driver-xPfC9vS5-py3.11/lib/python3.11/site-packages/recommonmark/parser.py:75: UserWarning: Container node skipped: type=document
>   warn("Container node skipped: type={0}".format(mdnode.t))
> ```

we have the same error on 3.x branch.

> 
> Visiting the `127.0.0.1:5500` directs me to `http://127.0.0.1:5500/master` which has 404 error. However pages like `http://127.0.0.1:5500/stable/index.html` or `http://127.0.0.1:5500/scylla-4.15.0.x/` remain functional with redirects to javadoc when clicking on API Documentation header too.
> 

i think this is due to the removed multiversion in previous reversion. and should be fixed.

> > > Branchest/Tags need an update: latest tag is 4.18.0.1
> > 
> > 
> > could you be more specific? which files shall i update?
> 
> I was guessing that we'd need to just add '4.18.0.1' to `TAGS` in docs/source/conf.py but that actually breaks the build with multiple warnings like:
> 

yeah, all 4.x tags are broken at this moment. i think the first step is to fix the 4.x branch.

> ```
> /tmp/tmpk4aq9ak4/970452ef5b72d76b4f783335fccabe57107d9735/docs/source/manual/core/pooling/index.md: WARNING: document isn't included in any toctree
> ```
> 
> So after all let's not do that for now.

tchaikov commented:
v3:

- revert the change to remove multiversion support.

tchaikov commented:
@Bouncheck i see you are still asking for changes. may i learn what exactly the changes i need to make?

Bouncheck commented:
I'm not getting redirected to javadoc when clicking on API Documentation on generated page (`make clean preview`).
But if that's intended for now I can approve anyway since this PR is a big improvement.
I haven't checked this on 3.x branch but if it's the same there it would make sense to me to approve now and have a separate fix for 3.x + port to 4.x later on.

tchaikov commented:
> I'm not getting redirected to javadoc when clicking on API Documentation on generated page (`make clean preview`). 

@Bouncheck to my best knowledge, this is indeed intended. see https://github.com/scylladb/java-driver/blob/9eedcaca09d1bb5bb87452389c1eb9cc1bd9645f/docs/source/conf.py#L174-L179

we redirect the api doc to `${version_name}/api/index.html`, and if `version_name` is 14.5.6, then the link points to `/14.5.6/api/index.html`, right under the root directory. i think this should work if the html is deployed to a web server.

> But if that's intended for now I can approve anyway since this PR is a big improvement. I haven't checked this on 3.x branch but if it's the same there it would make sense to me to approve now and have a separate fix for 3.x + port to 4.x later on.

i'd appreciate if you could help approve and merge this change. based on this change, we would be able to improve the 4.x document further.
